runtime/runtime/src/bandwidth_scheduler/mod.rs

```diff
@@ -38,55 +48,89 @@ pub fn run_bandwidth_scheduler(
+
+        let congestion_control = CongestionControl::new(
+            apply_state.config.congestion_control_config,
+            extended_congestion_info.congestion_info,
+            extended_congestion_info.missed_chunks_count,
+        );
+        let congestion_level = congestion_control.congestion_level();
+        let is_fully_congested = CongestionControl::is_fully_congested(congestion_level);
+
+        shards_status.insert(
+            *shard_id,
+            ShardStatus { last_chunk_missing, is_allowed_sender_shard, is_fully_congested },
+        );
+    }
 
+    // Prepare lists of sender and receiver shards.
+    // TODO(bandwidth_scheduler) - find a better way to get the sender and receiver shards.
+    // Maybe pass shard layout in ApplyState? That might also be needed for resharding.
+    // Taking all shards from the congestion info will be wrong for only one height during
+    // resharding, so it might be good enough for now, but it's not ideal.
     let mut all_shards = apply_state.congestion_info.all_shards();
     all_shards.sort();
     if all_shards.is_empty() {
-        all_shards = vec![ShardId::new(0)];
+        // In some tests there's no congestion info and the list of shards ends up empty.
+        // Pretend that there's only one default shard.
+        all_shards = vec![ShardUId::single_shard().shard_id()];
+        shards_status.insert(
+            all_shards[0],
+            ShardStatus {
+                last_chunk_missing: false,
+                is_allowed_sender_shard: true,
+                is_fully_congested: false,
+            },
+        );
     }
-    let num_shards =
-        NonZeroU64::new(all_shards.len().try_into().expect("Can't convert usize to u64"))
-            .expect("Number of shards can't be zero");
+    let sender_shards = &all_shards;
+    let receiver_shards = &all_shards;
 
-    let params = BandwidthSchedulerParams::new(num_shards, &apply_state.config);
+    // Calculate the current scheduler parameters.
+    let num_shards: u64 = std::cmp::max(sender_shards.len(), receiver_shards.len())
+        .try_into()
+        .expect("Can't convert usize to u64");
+    let params =
+        BandwidthSchedulerParams::new(NonZeroU64::new(num_shards).unwrap(), &apply_state.config);
 
-    let prev_block_hash = apply_state.prev_block_hash;
-    let bandwidth_requests = &apply_state.bandwidth_requests;
-    tracing::debug!(
-        target: "runtime",
-        ?prev_block_hash,
-        ?scheduler_state,
-        ?all_shards,
-        ?bandwidth_requests,
-        "Running bandwidth scheduler with inputs",
+    // Run the bandwidth scheduler algorithm.
+    let granted_bandwidth = BandwidthScheduler::run(
+        &sender_shards,
+        &receiver_shards,
+        &mut scheduler_state,
+        &params,
+        &apply_state.bandwidth_requests,
+        &shards_status,
+        apply_state.block_height,
```

Comment:
That's a good point. I think I initially did that, but there were some concerns about it. I wasn't sure if it's gonna work with optimistic block (/cc @Longarithm), and a lot of tests broke because they set random seed to zero and expect chunk application to have the same result as in reality ;-;
Block height isn't the best source of randomness, it's very predictable :/ Maybe `prev_block_hash` would be a bit better?

Reply:
I switched to using `prev_block_hash` for the scheduler Rng seed. It's more random than height and should be random enough while not causing any optimistic block issues.